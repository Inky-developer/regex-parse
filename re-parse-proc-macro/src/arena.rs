use std::fmt::Debug;
use std::hash::Hash;
use std::marker::PhantomData;
use std::ops::{Index, IndexMut};

#[derive(Debug)]
pub struct Arena<T> {
    nodes: Vec<T>,
}

impl<T> Arena<T> {
    pub fn add(&mut self, node: T) -> ArenaIndex<T> {
        let index = self.nodes.len();
        self.nodes.push(node);
        ArenaIndex::new(index)
    }
    
    pub fn iter(&self) -> impl Iterator<Item = ArenaIndex<T>> + use<'_, T> {
        self.nodes.iter().enumerate().map(|(i, _)| ArenaIndex::new(i))
    }
}

impl<T> Index<ArenaIndex<T>> for Arena<T> {
    type Output = T;

    fn index(&self, index: ArenaIndex<T>) -> &Self::Output {
        &self.nodes[index.index]
    }
}

impl<T> IndexMut<ArenaIndex<T>> for Arena<T> {
    fn index_mut(&mut self, index: ArenaIndex<T>) -> &mut Self::Output {
        &mut self.nodes[index.index]
    }
}

impl<T> Default for Arena<T> {
    fn default() -> Self {
        Self {
            nodes: Vec::default(),
        }
    }
}

pub struct ArenaIndex<T> {
    index: usize,
    _phantom: PhantomData<T>,
}

impl<T> ArenaIndex<T> {
    const fn new(index: usize) -> Self {
        Self {
            index,
            _phantom: PhantomData,
        }
    }
}

impl<T> Debug for ArenaIndex<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_tuple(&format!("ArenaIndex<{}>", std::any::type_name::<T>())).field(&self.index).finish()
    }
}

impl<T> Clone for ArenaIndex<T> {
    fn clone(&self) -> Self {
        Self {
            index: self.index,
            _phantom: PhantomData,
        }
    }
}

impl<T> Copy for ArenaIndex<T> {}

impl<T> PartialEq for ArenaIndex<T> {
    fn eq(&self, other: &Self) -> bool {
        self.index == other.index
    }
}

impl<T> Eq for ArenaIndex<T> {}

impl<T> Hash for ArenaIndex<T> {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.index.hash(state);
    }
}

impl<T> PartialOrd for ArenaIndex<T> {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        self.index.partial_cmp(&other.index)
    }
}

impl<T> Ord for ArenaIndex<T> {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.index.cmp(&other.index)
    }
}