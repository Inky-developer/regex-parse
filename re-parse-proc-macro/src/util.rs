use crate::Set;
use std::hash::Hash;

pub trait FloodFill {
    type Item;

    fn get_neighbors(&self, item: &Self::Item) -> impl Iterator<Item = Self::Item>;

    fn iter(&self, start: Self::Item) -> impl Iterator<Item = Self::Item>
    where
        Self: Sized,
        Self::Item: Eq + Hash + Clone,
    {
        let mut pending_nodes = Set::default();
        pending_nodes.insert(start);
        FloodFillIter {
            flood_fill: self,
            pending_nodes,
            visited_nodes: Set::default(),
        }
    }
}

struct FloodFillIter<'a, T: FloodFill> {
    flood_fill: &'a T,
    pending_nodes: Set<T::Item>,
    visited_nodes: Set<T::Item>,
}

impl<T> Iterator for FloodFillIter<'_, T>
where
    T: FloodFill,
    T::Item: Eq + Hash + Clone,
{
    type Item = T::Item;

    fn next(&mut self) -> Option<Self::Item> {
        let next = self.pending_nodes.iter().next()?;

        let next = next.clone();

        self.visited_nodes.insert(next.clone());
        self.pending_nodes.remove(&next);
        self.pending_nodes.extend(
            self.flood_fill
                .get_neighbors(&next)
                .filter(|neighbor| !self.visited_nodes.contains(neighbor)),
        );

        Some(next)
    }
}
