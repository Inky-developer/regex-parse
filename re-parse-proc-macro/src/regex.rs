use crate::arena::{Arena, ArenaIndex};
use crate::parser::{ParseError, RegexParser};
use crate::tokenizer::tokenize;
use std::fmt::{Debug, Display, Formatter, Write};

pub type RegexArena = Arena<RegexNode>;

pub type RegexNodeIndex = ArenaIndex<RegexNode>;

pub struct Regex {
    pub arena: RegexArena,
    pub root: RegexNodeIndex,
}

impl Regex {
    pub fn from_str(input: &str) -> Result<Self, ParseError> {
        RegexParser::parse(tokenize(input))
    }
}

impl Display for Regex {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        Display::fmt(
            &RegexDisplay {
                arena: &self.arena,
                node_idx: self.root,
            },
            f,
        )
    }
}

impl Debug for Regex {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        Debug::fmt(
            &RegexDisplay {
                arena: &self.arena,
                node_idx: self.root,
            },
            f,
        )
    }
}

#[derive(Debug)]
pub enum RegexNode {
    And(Vec<RegexNodeIndex>),
    Or(Vec<RegexNodeIndex>),
    Literal(RegexPattern),
    Variable(RegexVariable),
    ZeroOrOne(RegexNodeIndex),
    Many(RegexNodeIndex),
    OneOrMore(RegexNodeIndex),
}

#[derive(Debug, Clone, Copy)]
pub enum RegexPattern {
    Char(char),
    Range(char, char),
    AnyChar,
    /// Matches every character, except those that were explicitly specified.
    /// For example `(ABC|.)` (where `.` is [AnyChar]) matches the input `A`, because the `.`
    /// matched. If the `.` would be [AnyCharLazy], the regex would not match the input `A`, because
    /// the more specific patter `ABC` would take precedence.
    ///
    /// This is used for variables: `{var}` gets transformed into `.+`, where the `.` is lazy.
    /// The reason this is done is to make it possible to match anything at all.
    AnyCharLazy,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct RegexVariable {
    pub name: String,
    pub kind: VariableKind,
}

#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
pub enum VariableKind {
    Singular,
    Multiple,
}

pub struct RegexDisplay<'arena> {
    arena: &'arena RegexArena,
    node_idx: RegexNodeIndex,
}

impl RegexDisplay<'_> {
    fn node(&self, node_idx: RegexNodeIndex) -> Self {
        Self {
            arena: self.arena,
            node_idx,
        }
    }
}

impl Display for RegexDisplay<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let node = &self.arena[self.node_idx];

        match node {
            RegexNode::And(nodes) => {
                for node in nodes {
                    Display::fmt(&self.node(*node), f)?;
                }
            }
            RegexNode::Or(nodes) => {
                for (index, node) in nodes.iter().enumerate() {
                    Display::fmt(&self.node(*node), f)?;
                    if index + 1 < nodes.len() {
                        f.write_char('|')?;
                    }
                }
            }
            RegexNode::Literal(pat) => match pat {
                RegexPattern::Char(char) => f.write_char(*char)?,
                RegexPattern::Range(start, end) => write!(f, "{}-{}", start, end)?,
                RegexPattern::AnyChar | RegexPattern::AnyCharLazy => f.write_char('.')?,
            },
            RegexNode::Variable(RegexVariable { name, kind }) => match kind {
                VariableKind::Singular => write!(f, "{{{name}}}")?,
                VariableKind::Multiple => write!(f, "{{{name}*}}")?,
            },
            RegexNode::ZeroOrOne(node) => {
                Display::fmt(&self.node(*node), f)?;
                f.write_char('?')?;
            }
            RegexNode::Many(node) => {
                Display::fmt(&self.node(*node), f)?;
                f.write_char('*')?;
            }
            RegexNode::OneOrMore(node) => {
                Display::fmt(&self.node(*node), f)?;
                f.write_char('+')?;
            }
        }

        Ok(())
    }
}

impl Debug for RegexDisplay<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let node = &self.arena[self.node_idx];
        match node {
            RegexNode::And(nodes) => {
                let mut tuple = f.debug_tuple("And");
                for node in nodes {
                    tuple.field(&self.node(*node));
                }
                tuple.finish()?;
            }
            RegexNode::Or(nodes) => {
                let mut tuple = f.debug_tuple("Or");
                for node in nodes {
                    tuple.field(&self.node(*node));
                }
                tuple.finish()?;
            }
            RegexNode::Literal(literal) => f.debug_tuple("Literal").field(literal).finish()?,
            RegexNode::Variable(var) => f.debug_tuple("Variable").field(var).finish()?,
            RegexNode::ZeroOrOne(child) => f
                .debug_tuple("ZeroOrOne")
                .field(&self.node(*child))
                .finish()?,
            RegexNode::Many(child) => f.debug_tuple("Many").field(&self.node(*child)).finish()?,
            RegexNode::OneOrMore(child) => f
                .debug_tuple("OneOrMore")
                .field(&self.node(*child))
                .finish()?,
        }

        Ok(())
    }
}
