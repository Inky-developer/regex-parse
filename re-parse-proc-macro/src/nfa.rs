use crate::arena::{Arena, ArenaIndex};
use crate::regex::{Regex, RegexArena, RegexNode, RegexNodeIndex, RegexPattern, RegexVariable};
use crate::util::FloodFill;
use crate::Set;

pub type NfaArena = Arena<NfaNode>;
pub type NfaIndex = ArenaIndex<NfaNode>;

#[derive(Debug)]
pub struct Nfa {
    pub root: NfaIndex,
    pub nodes: NfaArena,
}

impl From<Regex> for Nfa {
    fn from(value: Regex) -> Self {
        let Regex { arena, root } = value;
        let mut nodes = NfaArena::default();
        let root_node = nodes.add(NfaNode::EPSILON);
        let target_node = convert_regex_node(&mut nodes, &arena, root, root_node);
        nodes[target_node].is_accepting = true;

        check_variables(&nodes);

        Nfa {
            nodes,
            root: root_node,
        }
    }
}

fn check_variables(nodes: &NfaArena) {
    let mut visited_variables = Set::default();
    for node in nodes.iter() {
        if let NfaNodeKind::Variable(RegexVariable { name, .. }) = &nodes[node].kind {
            if visited_variables.contains(name) {
                panic!(
                    "(TODO MAKE ERROR) Variable \"{}\" is already declared",
                    name
                );
            }
            visited_variables.insert(name.clone());
        }
    }
}

#[derive(Debug)]
pub struct NfaNode {
    pub edges: Vec<NfaIndex>,
    pub edge_kind: NfaEdge,
    pub kind: NfaNodeKind,
    pub is_accepting: bool,
}

impl NfaNode {
    pub const EPSILON: Self = Self {
        edges: Vec::new(),
        edge_kind: NfaEdge::Epsilon,
        kind: NfaNodeKind::Simple,
        is_accepting: false,
    };
}

#[derive(Debug)]
pub enum NfaNodeKind {
    Simple,
    Variable(RegexVariable),
}

#[derive(Debug)]
pub enum NfaEdge {
    Epsilon,
    Pattern(RegexPattern),
}

impl NfaEdge {
    pub fn is_epsilon(&self) -> bool {
        matches!(self, NfaEdge::Epsilon)
    }
}

fn convert_regex_node(
    arena: &mut NfaArena,
    regex_arena: &RegexArena,
    node: RegexNodeIndex,
    predecessor: NfaIndex,
) -> NfaIndex {
    let node = &regex_arena[node];
    match node {
        RegexNode::And(nodes) => {
            let mut last_node = predecessor;
            for node in nodes {
                let new_node = convert_regex_node(arena, regex_arena, *node, last_node);
                last_node = new_node;
            }
            last_node
        }
        RegexNode::Or(nodes) => {
            let target_node = arena.add(NfaNode::EPSILON);
            for node in nodes {
                let new_node = convert_regex_node(arena, regex_arena, *node, predecessor);
                arena.connect(new_node, target_node);
            }
            target_node
        }
        RegexNode::Literal(pattern) => arena.add_after(
            predecessor,
            NfaNode {
                edges: Vec::new(),
                edge_kind: NfaEdge::Pattern(*pattern),
                kind: NfaNodeKind::Simple,
                is_accepting: false,
            },
        ),
        RegexNode::Variable(var) => {
            let node = arena.add_after(
                predecessor,
                NfaNode {
                    edges: Vec::new(),
                    edge_kind: NfaEdge::Pattern(RegexPattern::AnyCharLazy),
                    kind: NfaNodeKind::Variable(var.clone()),
                    is_accepting: false,
                },
            );
            arena.connect(node, node);
            node
        }
        RegexNode::ZeroOrOne(child) => {
            let target_node = arena.add(NfaNode::EPSILON);
            arena.connect(predecessor, target_node);
            let new_node = convert_regex_node(arena, regex_arena, *child, predecessor);
            arena.connect(new_node, target_node);
            target_node
        }
        RegexNode::Many(child) => {
            let iteration_node = arena.add(NfaNode::EPSILON);
            arena.connect(predecessor, iteration_node);
            let target_node = arena.add(NfaNode::EPSILON);
            arena.connect(predecessor, target_node);
            let new_node = convert_regex_node(arena, regex_arena, *child, iteration_node);
            arena.connect(new_node, iteration_node);
            arena.connect(new_node, target_node);
            target_node
        }
        RegexNode::OneOrMore(child) => {
            let iteration_node = arena.add(NfaNode::EPSILON);
            arena.connect(predecessor, iteration_node);
            let target_node = arena.add(NfaNode::EPSILON);
            let new_node = convert_regex_node(arena, regex_arena, *child, iteration_node);
            arena.connect(new_node, iteration_node);
            arena.connect(new_node, target_node);
            target_node
        }
    }
}

impl NfaArena {
    fn connect(&mut self, source: NfaIndex, target: NfaIndex) {
        self[source].edges.push(target);
    }

    fn add_after(&mut self, source: NfaIndex, new_node: NfaNode) -> NfaIndex {
        let idx = self.add(new_node);
        self.connect(source, idx);
        idx
    }
}

impl FloodFill for Nfa {
    type Item = NfaIndex;

    fn get_neighbors(&self, item: &Self::Item) -> impl Iterator<Item = Self::Item> {
        self.nodes[*item].edges.iter().copied()
    }
}

#[cfg(test)]
mod tests {
    use crate::nfa::Nfa;
    use crate::parser::ParseError;
    use crate::regex::Regex;

    fn parse(source: &str) -> Result<Nfa, ParseError> {
        let regex = Regex::from_str(source)?;
        Ok(Nfa::from(regex))
    }

    #[test]
    fn test_regex_to_nfa() {
        insta::assert_debug_snapshot!(parse("A"));
        insta::assert_debug_snapshot!(parse("A|B|C"));
        insta::assert_debug_snapshot!(parse("A?b*c"));
        insta::assert_debug_snapshot!(parse(".{var}."));
        insta::assert_debug_snapshot!(parse(".+;"));
    }
}
