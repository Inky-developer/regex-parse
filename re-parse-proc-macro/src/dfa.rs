use crate::arena::{Arena, ArenaIndex};
use crate::nfa::{Nfa, NfaEdge, NfaIndex, NfaNodeKind};
use crate::regex::{RegexPattern, RegexVariable};
use crate::util::FloodFill;
use crate::{Map, Set};

pub type DfaArena = Arena<DfaNode>;
pub type DfaIndex = ArenaIndex<DfaNode>;

#[derive(Debug)]
pub struct Dfa {
    pub root: DfaIndex,
    pub nodes: DfaArena,
}

impl Dfa {
    pub fn iter(&self) -> impl Iterator<Item = DfaIndex> + use<'_> {
        <Self as FloodFill>::iter(self, self.root)
    }
}

impl From<Nfa> for Dfa {
    fn from(nfa: Nfa) -> Self {
        let mut builder = DfaBuilder::default();
        let hull = calculate_epsilon_hull(&nfa);
        let root_group = hull[&nfa.root].clone();
        builder.pending_nodes.extend(hull.into_values());

        while let Some(group) = builder.pending_nodes.iter().next() {
            let group = group.clone();
            builder.pending_nodes.remove(&group);

            builder.compute_group(&nfa, group);
        }

        let root = builder.nfa_to_dfa[&root_group];
        Dfa {
            root,
            nodes: builder.nodes,
        }
    }
}

impl FloodFill for Dfa {
    type Item = DfaIndex;

    fn get_neighbors(&self, item: &Self::Item) -> impl Iterator<Item = Self::Item> {
        let edges = &self.nodes[*item].edges;
        edges
            .default
            .iter()
            .copied()
            .chain(edges.edges.values().copied())
    }
}

#[derive(Debug, Default)]
pub struct DfaBuilder {
    nodes: DfaArena,
    nfa_to_dfa: Map<Vec<NfaIndex>, DfaIndex>,
    pending_nodes: Set<Vec<NfaIndex>>,
}

impl DfaBuilder {
    fn insert(&mut self, key: Vec<NfaIndex>, node: DfaNode) -> DfaIndex {
        if let Some(idx) = self.nfa_to_dfa.get(&key) {
            self.nodes[*idx] = node;
            return *idx;
        }

        let idx = self.nodes.add(node);
        self.nfa_to_dfa.insert(key, idx);
        idx
    }

    fn entry(&mut self, key: Vec<NfaIndex>) -> DfaIndex {
        if let Some(idx) = self.nfa_to_dfa.get(&key) {
            return *idx;
        }

        let node = DfaNode::default();
        self.pending_nodes.insert(key.clone());
        self.insert(key, node)
    }

    fn compute_group(&mut self, nfa: &Nfa, group: Vec<NfaIndex>) {
        let edges = DfaEdges::from_nfa_group(self, nfa, &group);
        let is_accepting = group
            .iter()
            .copied()
            .any(|nfa_idx| nfa.nodes[nfa_idx].is_accepting);
        let variable = self.compute_group_variable(nfa, &group);

        self.insert(
            group,
            DfaNode {
                is_accepting,
                variable,
                edges,
            },
        );
    }

    fn compute_group_variable(&self, nfa: &Nfa, group: &[NfaIndex]) -> Option<RegexVariable> {
        let mut variable = None;

        for nfa_idx in group.iter().copied() {
            let NfaNodeKind::Variable(var) = &nfa.nodes[nfa_idx].kind else {
                continue;
            };

            match variable {
                None => variable = Some(var.clone()),
                Some(RegexVariable {
                    name: other_var, ..
                }) => panic!(
                    "(TODO THROW ERROR) ambiguous variables: {other_var} collides with {}",
                    &var.name
                ),
            }
        }

        variable
    }
}

fn get_non_epsilon_edges(nfa: &Nfa, group: &[NfaIndex]) -> Vec<(RegexPattern, NfaIndex)> {
    let mut edges: Vec<(RegexPattern, NfaIndex)> = Vec::new();
    for node_idx in group {
        let node = &nfa.nodes[*node_idx];
        for edge_idx in &node.edges {
            let edge = &nfa.nodes[*edge_idx];
            if let NfaEdge::Pattern(pattern) = &edge.edge_kind {
                edges.push((*pattern, *edge_idx))
            }
        }
    }
    edges
}

fn calculate_epsilon_hull(nfa: &Nfa) -> Map<NfaIndex, Vec<NfaIndex>> {
    let mut hull = Map::default();

    for idx in nfa.iter() {
        hull.insert(idx, get_connected_nodes(nfa, idx));
    }

    hull
}

fn expand_group(nfa: &Nfa, group: &[NfaIndex]) -> Vec<NfaIndex> {
    let mut nodes = Set::default();
    for idx in group.iter().copied() {
        nodes.extend(get_connected_nodes(nfa, idx));
    }

    let mut result = nodes.into_iter().collect::<Vec<_>>();
    result.sort();
    result
}

fn get_connected_nodes(nfa: &Nfa, idx: NfaIndex) -> Vec<NfaIndex> {
    let mut nodes: Set<NfaIndex> = Set::default();
    let mut pending_nodes: Set<NfaIndex> = Set::default();

    pending_nodes.insert(idx);
    while let Some(node) = pending_nodes.iter().copied().next() {
        pending_nodes.remove(&node);
        nodes.insert(node);

        pending_nodes.extend(
            nfa.nodes[node]
                .edges
                .iter()
                .copied()
                .filter(|edge| nfa.nodes[*edge].edge_kind.is_epsilon()),
        )
    }

    let mut result: Vec<NfaIndex> = nodes.into_iter().collect();
    result.sort();
    result
}

#[derive(Debug, Default)]
pub struct DfaNode {
    pub is_accepting: bool,
    pub variable: Option<RegexVariable>,
    pub edges: DfaEdges,
}

#[derive(Debug, Default)]
pub struct DfaEdges {
    pub default: Option<DfaIndex>,
    pub edges: Map<char, DfaIndex>,
}

impl DfaEdges {
    fn from_nfa_group(dfa: &mut DfaBuilder, nfa: &Nfa, group: &[NfaIndex]) -> Self {
        let edges = get_non_epsilon_edges(nfa, group);

        let mut default_edges: Vec<NfaIndex> = Vec::new();
        let mut lazy_default_edges: Vec<NfaIndex> = Vec::new();

        let mut edge_map: Map<char, Vec<NfaIndex>> = Map::default();
        for (edge_pattern, target_idx) in edges.iter().copied() {
            match edge_pattern {
                RegexPattern::Char(char) => edge_map.entry(char).or_default().push(target_idx),
                RegexPattern::Range(start, end) => {
                    for char in start..=end {
                        edge_map.entry(char).or_default().push(target_idx);
                    }
                }
                RegexPattern::AnyChar => default_edges.push(target_idx),
                RegexPattern::AnyCharLazy => lazy_default_edges.push(target_idx),
            }
        }

        // Since a default edge can be any char, it also has to be added to each value in the edge map now.
        for targets in edge_map.values_mut() {
            targets.extend(default_edges.iter().copied());
            targets.sort_unstable();
            targets.dedup();
        }

        // If there is a default_edge, it will overwrite the lazy-default edge completely.
        if default_edges.is_empty() {
            default_edges = lazy_default_edges;
        }
        default_edges.sort_unstable();
        default_edges.dedup();

        let default_edge_idx = if default_edges.is_empty() {
            None
        } else {
            Some(dfa.entry(expand_group(nfa, &default_edges)))
        };
        let edge_indices = edge_map
            .into_iter()
            .map(|(key, value)| (key, dfa.entry(expand_group(nfa, &value))))
            .collect();
        DfaEdges {
            default: default_edge_idx,
            edges: edge_indices,
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::dfa::Dfa;
    use crate::nfa::Nfa;
    use crate::parser::ParseError;
    use crate::regex::Regex;

    fn parse(input: &str) -> Result<Dfa, ParseError> {
        let regex = Regex::from_str(input)?;
        let nfa = Nfa::from(regex);
        Ok(Dfa::from(nfa))
    }

    #[test]
    fn test_nfa_to_dfa() {
        insta::assert_debug_snapshot!(parse("A"));
        insta::assert_debug_snapshot!(parse("AB"));
        insta::assert_debug_snapshot!(parse("A?B"));
        insta::assert_debug_snapshot!(parse("A?A"));
        insta::assert_debug_snapshot!(parse("A?b*c"));
        insta::assert_debug_snapshot!(parse("{foo}"));
        insta::assert_debug_snapshot!(parse("A{foo}B+{bar}"));
        insta::assert_debug_snapshot!(parse("[a-e]"));
        insta::assert_debug_snapshot!(parse(".{var}."));
    }

    #[test]
    #[should_panic(expected = "(TODO THROW ERROR) ambiguous variables: foo collides with bar")]
    fn test_nfa_to_dfa_ambiguous_variable() {
        parse("A{foo}B?{bar}").unwrap();
    }
}
