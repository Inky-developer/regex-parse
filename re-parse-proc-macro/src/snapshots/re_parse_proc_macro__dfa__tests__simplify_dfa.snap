---
source: re-parse-proc-macro/src/dfa.rs
expression: "parse(\"([abc]\\\\s*)*\")"
snapshot_kind: text
---
Ok(
    Dfa {
        root: ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
            3,
        ),
        nodes: Arena {
            nodes: [
                DfaNode {
                    is_accepting: true,
                    variable: None,
                    edges: DfaEdges {
                        default: None,
                        edges: {
                            ' ': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\r': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'b': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\n': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'a': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\t': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'c': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                        },
                    },
                },
                DfaNode {
                    is_accepting: true,
                    variable: None,
                    edges: DfaEdges {
                        default: None,
                        edges: {
                            ' ': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\r': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'b': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\n': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'a': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\t': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'c': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                        },
                    },
                },
                DfaNode {
                    is_accepting: true,
                    variable: None,
                    edges: DfaEdges {
                        default: None,
                        edges: {
                            ' ': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\r': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'b': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\n': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'a': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\t': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'c': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                        },
                    },
                },
                DfaNode {
                    is_accepting: true,
                    variable: None,
                    edges: DfaEdges {
                        default: None,
                        edges: {
                            'a': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'b': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'c': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                        },
                    },
                },
                DfaNode {
                    is_accepting: true,
                    variable: None,
                    edges: DfaEdges {
                        default: None,
                        edges: {
                            ' ': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\r': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'b': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\n': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'a': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\t': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'c': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                        },
                    },
                },
                DfaNode {
                    is_accepting: true,
                    variable: None,
                    edges: DfaEdges {
                        default: None,
                        edges: {
                            ' ': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\r': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'b': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\n': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'a': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\t': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'c': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                        },
                    },
                },
                DfaNode {
                    is_accepting: true,
                    variable: None,
                    edges: DfaEdges {
                        default: None,
                        edges: {
                            ' ': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\r': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'b': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\n': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'a': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\t': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'c': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                        },
                    },
                },
                DfaNode {
                    is_accepting: true,
                    variable: None,
                    edges: DfaEdges {
                        default: None,
                        edges: {
                            ' ': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\r': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'b': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\n': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'a': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            '\t': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                            'c': ArenaIndex<re_parse_proc_macro::dfa::DfaNode>(
                                0,
                            ),
                        },
                    },
                },
            ],
        },
    },
)
